MetaMask/metamask-extension
# Fix Sentry deduplication of events that were never sent

The Sentry `Dedupe` integration has been filtering out our events, even when they were never sent due to our `beforeSend` handler. It was wrongly identifying them as duplicates because it has no knowledge of `beforeSend` or whether they were actually sent or not.

To resolve this, the filtering we were doing in `beforeSend` has been moved to a Sentry integration. This integration is installed ahead of the `Dedupe` integration, so `Dedupe` should never find out about any events that we filter out, and thus will never consider them as sent when they were not.

Fixes #15666

## Manual Testing Steps

1. Setup errors for testing
  Before testing, ensure you know of a reliable way to trigger a Sentry error. If not, introduce an error somewhere in the codebase for testing purposes. We will want to test errors thrown in both the background and the UI, so we'll want a way to test both.
Typically my solution for the background error is to add `setImmediate(() => { throw new Error('test') })` in one of the setters in the preferences controller, as those are easy to trigger. The `setImmediate` ensures that the error stays in the background context, rather than being returned to the UI.
For the UI function, I throw an error in the `render` function for a page.
2. Set `SENTRY_DSN_DEV` environment variable or `.metamaskrc` entry to the Sentry DSN for either the test MetaMask account or your personal Sentry account. I've found personal accounts to be more reliable for testing, so that's what I'd recommend.
3. Comment out [this line](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/setupSentry.js#L74) if testing with a dev build
We will want to test both prod and dev builds. That line disables Sentry on dev builds.
4. Build and install MetaMask, making sure to **disable** MetaMetrics during onboarding.
5. Open the background dev console to the network tab
6. Trigger the background error, and see that there is no corresponding network request.
7. Open the fullscreen UI, and open the dev console for that page to the network tab.
8. Trigger the UI error, and see that there is no corresponding network request.
9. Enable MetaMetrics
10. Trigger the background error, and ensure there is a corresponding Sentry network request containing the error report.
11. Trigger the UI error, and ensure there is a corresponding Sentry network request containing the error report.
12. Check the Sentry dashboard corresponding to the `SENTRY_DSN_DEV` you set, and ensure both errors are shown.

## Pre-Merge Checklist

  - This seems quite difficult to test. We don't have any test coverage of Sentry. Created a ticket to track this: #15681

### + If there are functional changes:
